File: src/main/java/de/mm/fileshuffler/Chooser.java
package de.mm.fileshuffler;

import de.mm.fileshuffler.filter.MovieFilter;
import de.mm.fileshuffler.filter.MusicFilter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.imageio.ImageIO;
import javax.swing.*;
import java.awt.*;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.FileFilter;
import java.io.IOException;
import java.io.InputStream;
import java.util.*;
import java.util.List;

public class Chooser {
	private static final Logger LOGGER = LoggerFactory.getLogger(Chooser.class);

	private Map<File, Boolean> paths;
	private FileFilter filter;
	private Queue<File> content;
	private boolean isScanningNecessary;
	private TrayIcon trayIcon;
	private Menu pathChooser;

	public Chooser() {
		this.paths = new HashMap<>();
		this.paths.put(new File(System.getProperty("user.dir")), true);
		this.filter = new MovieFilter();
		this.isScanningNecessary = true;
		initSystemTray();
	}

	private void shuffle() {
		LOGGER.debug("shuffle with scanning: '{}'", isScanningNecessary);
		if (isScanningNecessary) {
			List<File> files = scanFolders(paths, filter);
			isScanningNecessary = false;
			Collections.shuffle(files);
			content = new LinkedList<>(files);
			handleFile(content.poll());
		}

		if (content.isEmpty()) {
			JOptionPane.showMessageDialog(null, "Nichts passendes gefunden :( (Filter?)", ":(", JOptionPane.INFORMATION_MESSAGE);
		}
	}

	private List<File> scanFolders(Map<File, Boolean> paths, FileFilter filter) {
		List<File> files = new ArrayList<>();
		for (File path : paths.keySet()) {
			if (paths.get(path)) {
				files.addAll(getFiles(path, filter));
			}
		}
		return files;
	}

	private List<File> getFiles(File path, FileFilter filter) {
		LOGGER.debug("getFiles from {}", path.getAbsolutePath());
		List<File> content = new ArrayList<>();
		for (File file : path.listFiles(filter)) {
			if (file.isDirectory()) {
				content.addAll(getFiles(file, filter));
			} else {
				content.add(file);
			}
		}
		return content;
	}

	private void handleFile(File f) {
		LOGGER.debug("handleFile: '{}'", f.getAbsolutePath());
		if (Desktop.isDesktopSupported()) {
			try {
				Desktop.getDesktop().open(f);
				displaySystemTrayMessage(f.getName());
			} catch (IOException e) {
				LOGGER.error("handleFile '{}' konnte nicht geöffnent werden", e);
				JOptionPane.showMessageDialog(null, "Fehler beim Handeln\n " + e.getMessage(), "Error", JOptionPane.ERROR_MESSAGE);
			}
		} else {
			LOGGER.warn("handleFile '{}' konnte nicht geöffnent werden (Desktop not supported", f);
			JOptionPane.showMessageDialog(null, "File:\n" + f + "\n wurde gewürfelt, konnte aber nicht geöffnent werden", "Error", JOptionPane.WARNING_MESSAGE);
		}
	}
	
	/*
	 * Path
	 */

	private void addPath() {
		LOGGER.debug("addPath");
		JFileChooser chooser = new JFileChooser(paths.keySet().iterator().next());
		chooser.setFileSelectionMode(JFileChooser.DIRECTORIES_ONLY);
		int returnVal = chooser.showOpenDialog(null);
		if (returnVal == JFileChooser.APPROVE_OPTION) {
			File path = chooser.getSelectedFile();
			LOGGER.debug("addPath '{}'", path);
			this.paths.put(path, true);
			this.isScanningNecessary = true;
			displaySystemTrayMessage("Path: " + path.getPath());
			updateChooserIconBoxes();
		}
	}

	private void updateChooserIconBoxes() {
		LOGGER.debug("updateChooserIconBoxes");
		pathChooser.removeAll();
		for (File path : paths.keySet()) {
			CheckboxMenuItem checkboxMenuItem = new CheckboxMenuItem(path.getName(), paths.get(path));
			checkboxMenuItem.addItemListener(e -> togglePath(path, checkboxMenuItem.getState()));
			pathChooser.add(checkboxMenuItem);
		}
	}

	private void togglePath(File path, boolean shouldAdd) {
		LOGGER.debug("togglePath {} {}", path, shouldAdd);
		isScanningNecessary = true;
		if (shouldAdd) {
			paths.put(path, shouldAdd);
			displaySystemTrayMessage("add path: " + path);
		} else {
			paths.remove(path);
			displaySystemTrayMessage("remove path: " + path);
		}
	}

	private void setFilter(FileFilter filter) {
		LOGGER.debug("setFilter to '{}'", filter);
		this.filter = filter;
		this.isScanningNecessary = true;
		displaySystemTrayMessage("Filter: " + filter.toString());
	}
	
	/*
	 * SystemTray
	 */

	private void initSystemTray() {
		LOGGER.debug("initSystemTray");
		BufferedImage trayIconImage;
		try {
			InputStream s = Thread.currentThread().getContextClassLoader().getResourceAsStream("images/dice.png");
			trayIconImage = ImageIO.read(s);
		} catch (IOException e) {
			LOGGER.error("error while reading dice image", e);
			return;
		}
		int trayIconWidth = new TrayIcon(trayIconImage).getSize().width;
		PopupMenu popup = new PopupMenu();

		MenuItem playItem = new MenuItem("Play a new one");
		playItem.addActionListener(e -> shuffle());

		// path

		MenuItem pathAdder = new MenuItem("add Path");
		pathAdder.addActionListener(e -> addPath());

		pathChooser = new Menu("choose Path");
		updateChooserIconBoxes();

		// filter

		MenuItem movieItem = new MenuItem("Movies");
		movieItem.addActionListener(e -> setFilter(new MovieFilter()));

		MenuItem musicItem = new MenuItem("Music");
		musicItem.addActionListener(e -> setFilter(new MusicFilter()));

		Menu filterChooser = new Menu("choose FilterTyp");
		filterChooser.add(movieItem);
		filterChooser.add(musicItem);

		// exit

		MenuItem exitItem = new MenuItem("Exit");
		exitItem.addActionListener(e -> exitProgram());

		popup.add(playItem);
		popup.addSeparator();
		popup.add(pathAdder);
		popup.add(pathChooser);
		popup.add(filterChooser);
		popup.addSeparator();
		popup.add(exitItem);

		trayIcon = new TrayIcon(trayIconImage.getScaledInstance(trayIconWidth, -1, Image.SCALE_SMOOTH), "FileShuffler", popup);
		try {
			SystemTray.getSystemTray().add(trayIcon);
			displaySystemTrayMessage("ready for orders");
		} catch (AWTException ignored) {
		}

	}

	private void displaySystemTrayMessage(String message) {
		LOGGER.debug("displaySystemTrayMessage: '{}'", message);
		if (trayIcon != null) {
			trayIcon.displayMessage("FileShuffler", message, TrayIcon.MessageType.INFO);
		}
	}

	private void clearSystemTray() {
		LOGGER.debug("clearSystemTray");
		if (!SystemTray.isSupported()) {
			return;
		}
		SystemTray.getSystemTray().remove(trayIcon);
	}

	/*
	 * exit
	 */

	private void exitProgram() {
		LOGGER.debug("exitProgram");
		clearSystemTray();
		System.exit(0);
	}

}

File: src/main/kotlin/de/mm/fileshuffler/Chooser.kt
package de.mm.fileshuffler

import de.mm.fileshuffler.filter.ExtensionFilter
import de.mm.fileshuffler.service.PreferenceService
import de.mm.fileshuffler.service.RecentFolderService
import org.slf4j.LoggerFactory
import java.awt.*
import java.awt.image.BufferedImage
import java.io.File
import java.io.FileFilter
import java.io.IOException
import javax.imageio.ImageIO
import javax.swing.JFileChooser
import javax.swing.JOptionPane

class Chooser {
	private val paths: MutableMap<File, Boolean> = mutableMapOf()
	private val movieFilter: ExtensionFilter
	private val musicFilter: ExtensionFilter
	private val preferenceService = PreferenceService()
	private var filter: FileFilter
	private var isScanningNecessary: Boolean
	private var isNotificationEnabled: Boolean
	private var mode: Mode
	private var trayIcon: TrayIcon? = null
	private lateinit var pathChoicer: Menu
	private lateinit var modeItem: MenuItem

	init {
		this.isNotificationEnabled = preferenceService.isNotificationEnabled()
		this.movieFilter = ExtensionFilter(preferenceService.getMovieExtensions())
		this.musicFilter = ExtensionFilter(preferenceService.getMusicExtensions())
		this.filter = this.movieFilter
		this.paths.putAll(RecentFolderService.getRecentFolders().map { Pair(it, true) })
		this.isScanningNecessary = true
		this.mode = Mode.RANDOM
		initSystemTray(this.mode)
	}

	/*
	 * scan
	 */

	private fun shuffle() {
		LOGGER.debug("shuffle with scanning: '{}'", isScanningNecessary)
		if (isScanningNecessary) {
			mode.init(paths, filter)
			isScanningNecessary = false
		}

		val file = mode.pop()
		if (file == null) {
			displayDialog("Nichts passendes gefunden :( (Filter?)")
		} else {
			handleFile(file)
		}
	}

	private fun handleFile(f: File) {
		LOGGER.debug("handleFile: '{}'", f.absolutePath)
		if (Desktop.isDesktopSupported()) {
			try {
				Desktop.getDesktop().open(f)
				displaySystemTrayMessage(f.name)
			} catch (e: IOException) {
				LOGGER.error("handleFile '{}' konnte nicht geöffnent werden", e)
				displayDialog("Fehler beim Handeln\n: ${e.message}", JOptionPane.ERROR_MESSAGE)
			}
		} else {
			LOGGER.warn("handleFile '{}' konnte nicht geöffnent werden (Desktop not supported", f)
			displayDialog("File:\n$f\n wurde gewürfelt, konnte aber nicht geöffnent werden", JOptionPane.WARNING_MESSAGE)
		}
	}

	/*
	 * Path
	 */

	private fun addPath() {
		LOGGER.debug("addPath")
		val chooser = JFileChooser()
		chooser.fileSelectionMode = JFileChooser.DIRECTORIES_ONLY
		val returnVal = chooser.showOpenDialog(null)
		if (returnVal == JFileChooser.APPROVE_OPTION) {
			val path = chooser.selectedFile
			LOGGER.debug("addPath '{}'", path)
			this.paths.put(path, true)
			this.isScanningNecessary = true
			displaySystemTrayMessage("Path: " + path.path)
			updateChooserIconBoxes()
		}
	}

	private fun updateChooserIconBoxes() {
		LOGGER.debug("updateChooserIconBoxes")
		pathChoicer.removeAll()
		paths.keys.forEach {
			val checkboxMenuItem = CheckboxMenuItem(it.name, paths[it]!!)
			checkboxMenuItem.addItemListener { _ -> togglePath(it, checkboxMenuItem.state) }
			pathChoicer.add(checkboxMenuItem)
		}
	}

	private fun togglePath(path: File, shouldAdd: Boolean) {
		LOGGER.debug("togglePath {} {}", path, shouldAdd)
		isScanningNecessary = true
		if (shouldAdd) {
			paths.put(path, shouldAdd)
			displaySystemTrayMessage("add path: " + path)
		} else {
			paths.remove(path)
			displaySystemTrayMessage("remove path: " + path)
		}
	}

	private fun setFilter(filter: FileFilter) {
		LOGGER.debug("setFilter to '{}'", filter)
		this.filter = filter
		this.isScanningNecessary = true
		displaySystemTrayMessage("Filter: " + filter.toString())
	}

	/*
	 * SystemTray
	 */

	private fun initSystemTray(mode: Mode) {
		LOGGER.debug("initSystemTray")

		// play
		val playItem = MenuItem("Play a new one").apply {
			addActionListener { shuffle() }
		}

		// mode
		modeItem = MenuItem().apply {
			addActionListener { toggleMode(mode.other()) }
			label = when (mode) {
				Mode.QUEUE -> "Mode (use random))"
				Mode.RANDOM -> "Mode (use queue)"
			}
		}

		// path
		pathChoicer = Menu("choose")
		updateChooserIconBoxes()

		val pathChooser = Menu("Path").apply {
			add(MenuItem("add").apply {
				addActionListener { addPath() }
			})
			add(pathChoicer)
		}

		// filter
		val filterChooser = Menu("FilterTyp").apply {
			add(MenuItem("Movies").apply {
				addActionListener { setFilter(movieFilter) }
			})
			add(MenuItem("Music").apply {
				addActionListener { setFilter(musicFilter) }
			})
		}

		// queue
		val queueChooser = Menu("Queue").apply {
			add(MenuItem("Reload").apply {
				addActionListener {
					LOGGER.debug("refill queue")
					mode.reload(paths, filter)
				}
			})
		}

		// settings
		val settingsChooser = Menu("Settings").apply {
			add(CheckboxMenuItem("enable messages", isNotificationEnabled).apply {
				addItemListener { _ -> toggleNotification(state) }
			})
		}

		// exit
		val exitItem = MenuItem("Exit").apply {
			addActionListener { exitProgram() }
		}

		val popup = PopupMenu().apply {
			add(playItem)
			addSeparator()
			add(modeItem)
			if (mode === Mode.QUEUE) {
				add(queueChooser)
			}
			add(pathChooser)
			add(filterChooser)
			addSeparator()
			add(settingsChooser)
			addSeparator()
			add(exitItem)
		}

		val diceImage = Thread.currentThread().contextClassLoader.getResourceAsStream("images/dice.png")
		val trayIconImage = ImageIO.read(diceImage)
		if (trayIcon != null){
			SystemTray.getSystemTray().remove(trayIcon)
		}
		trayIcon = TrayIcon(trayIconImage.getScaledInstance(TrayIcon(trayIconImage).size.width, -1, Image.SCALE_SMOOTH), "FileShuffler", popup)
		SystemTray.getSystemTray().add(trayIcon)
		displaySystemTrayMessage("ready for orders")

	}

	private fun toggleNotification(state: Boolean) {
		LOGGER.debug("toggleNotification: '{}'", state)
		isNotificationEnabled = state
		preferenceService.setNotificationEnabled(state)
	}

	private fun toggleMode(mode: Mode) {
		LOGGER.debug("toggleMode to: '{}'", mode)
		this.mode = mode
		modeItem.label = when (mode) {
			Mode.QUEUE -> "Mode (use random))"
			Mode.RANDOM -> "Mode (use queue)"
		}
		initSystemTray(mode)
	}

	private fun displaySystemTrayMessage(message: String) {
		LOGGER.debug("displaySystemTrayMessage: '{}' {}", message, isNotificationEnabled)
		if (isNotificationEnabled) {
			trayIcon?.displayMessage("FileShuffler", message, TrayIcon.MessageType.INFO)
		}
	}

	private fun displayDialog(message: String, mode: Int = JOptionPane.INFORMATION_MESSAGE) {
		LOGGER.debug("displayDialog '{}'", message)
		JOptionPane.showMessageDialog(null, message, "Error", mode)
	}

	/*
	 * exit
	 */

	private fun exitProgram() {
		LOGGER.debug("exitProgram")
		mode.shutdown()
		preferenceService.storePreferences()
		RecentFolderService.storeRecentFolders(paths.keys)
		SystemTray.getSystemTray().remove(trayIcon)
		System.exit(0)
	}

	companion object {
		private val LOGGER = LoggerFactory.getLogger(Chooser::class.java)
	}

}